File: frontend/types/workspace.ts
export interface ResearchSourceRead {
    page_id: string;
    search_result_id: string;
    title: string;
    url: string;
    domain: string;
    source_type: string;
    status: string;
    analysis_status: string;
    quality_score: number;
    credibility_score: number;
    fetch_duration_ms?: number | null;
    analysis_duration_ms?: number | null;
    word_count: number;
    token_count: number;
    extraction_quality_score: number;
    summary?: string | null;
    key_claims: string[];
    entities: string[];
    relationships: string[];
    research_relevance: string;
    created_at: string;
}

export interface ResearchSourcesResponse {
    session_id: string;
    sources: ResearchSourceRead[];
}

export interface KnowledgeNodeRead {
    id: string;
    session_id: string;
    concept: string;
    description: string;
    confidence: number;
    source_count: number;
    created_at: string;
}

export interface KnowledgeEdgeRead {
    id: string;
    session_id: string;
    source_node: string;
    target_node: string;
    relationship: string;
    created_at: string;
}

export interface KnowledgeGraphResponse {
    nodes: KnowledgeNodeRead[];
    edges: KnowledgeEdgeRead[];
}

File: app/models/workspace.py
from datetime import datetime
from uuid import UUID
from typing import List, Optional

from sqlmodel import SQLModel, Field


class ResearchSourceRead(SQLModel):
    """
    Read model for a source used during research.
    """
    page_id: UUID
    search_result_id: UUID
    title: str
    url: str
    domain: str
    source_type: str
    status: str
    analysis_status: str
    quality_score: float
    credibility_score: float
    fetch_duration_ms: Optional[float] = None
    analysis_duration_ms: Optional[float] = None
    word_count: int = 0
    token_count: int = 0
    extraction_quality_score: float = 0.0
    summary: Optional[str] = None
    key_claims: List[str] = Field(default_factory=list)
    entities: List[str] = Field(default_factory=list)
    relationships: List[str] = Field(default_factory=list)
    research_relevance: str = "Unknown"
    created_at: datetime


class ResearchSourcesResponse(SQLModel):
    """
    Response model for all sources used in a session.
    """
    session_id: UUID
    sources: List[ResearchSourceRead] = Field(default_factory=list)